### PR TITLE
PERF: introduce fn-local modification trackers

### PR DESCRIPTION
**Pull request by vlad20012**
Now we will invalidate type inference cache only if a current function or program structure was changed. I.e. if we type in some function, inferred types of other functions will not be flushed.

Should fix most of "high CPU usage" issues

### CODE REVIEW

**Comment by vlad20012** [comment]
Hmm, carg-fmt failed on nightly

File: src/main/kotlin/org/rust/lang/core/psi/RsPsiManager.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            updateModificationCount(child ?: event.parent)
+        }
+    }
+
+    private fun updateModificationCount(psi: PsiElement) {
+        // We find a nearest parent item or macro call (because macro call can produce items)
+        // If found item implements RsModificationTrackerOwner, we increment it's own
+        // modification counter. Otherwise we increment global modification counter.
+        //
+        // So, if we changing something inside a functions and it is not an item,
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/psi/RsPsiManager.kt]: Let's better write something like this:
"So, if something is changed inside a function except an item, we will only increment the function local modification counter."

File: src/main/kotlin/org/rust/lang/core/psi/RsPsiManager.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        // we incrementing the local modification counter of the function only.
+        //
+        // It may not be intuitive that if we changing an item inside a function,
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/psi/RsPsiManager.kt]: changing -> change

File: src/main/kotlin/org/rust/lang/core/psi/ext/PsiElement.kt [hunk 3]
```diff
@@ -96,3 +96,6 @@ fun PsiElement?.getPrevNonCommentSibling(): PsiElement? =
  */
 fun PsiElement?.getNextNonCommentSibling(): PsiElement? =
     PsiTreeUtil.skipSiblingsForward(this, PsiWhiteSpace::class.java, PsiComment::class.java)
+
+fun RsElement.isParentOf(child: PsiElement): Boolean =
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/psi/ext/PsiElement.kt]: `isAncestorOf`? Or change `ancestors` to `parents` :)

File: src/test/kotlin/org/rustSlowTests/RsHighlightingPerformanceTest.kt [hunk 4]
```diff
@@ -67,6 +69,20 @@ class RsHighlightingPerformanceTest : RustWithToolchainTestBase() {
...
+
+        val stmt = myFixture.file.descendantsOfType<RsFunction>().asSequence()
+            .mapNotNull { it.block?.stmtList?.lastOrNull() }.first()
+        myFixture.editor.caretModel.moveToOffset(stmt.textOffset)
+        myFixture.type("2+2;")
+        PsiDocumentManager.getInstance(project).commitAllDocuments() // process PSI modification events
+
+        timings.measure("resolve_after_typing") {
+            refs.forEach { it.reference.resolve() }
+        }
```
Comment (matklad) [inline src/test/kotlin/org/rustSlowTests/RsHighlightingPerformanceTest.kt]: :+1:

**Pull request merged**